man/neat.1
.\" Manpage for neat.
.TH man 1 "5 June 2016" "2.4" "neat man page"
.SH NAME
\fBneat\fR \- nebular empirical analysis tool
.SH SYNOPSIS
\fBneat\fR [option1] [value1] ...
.SH DESCRIPTION
\fBneat\fR carries out a comprehensive empirical analysis on a list of nebular emission line fluxes. If uncertainties are required, it evaluates them using a Monte Carlo approach.  \fBneat\fR's philosophy is that abundance determinations should be as objective as possible, and so user input should be minimised. A number of choices can be made by the user, but the default options should yield meaningful results.
.SH OPTIONS

.TP
.B \-c [VALUE]
The logarithmic extinction at H beta.  By default, this is calculated from the Balmer line ratios.

.TP
.B \-cf, \-\-configuration-file [FILE]
Since version 2.0, NEAT's analysis scheme is specified in a configuration file which is read in at run time, rather than being hard coded.  If this option is omitted the default configuration file is used, which is intended to be suitable for any line list and should only really need changing if particular lines need to be excluded from the analysis.  The configuration file contains a series of weights to be applied when calculating diagnostics and abundances, and the format of the default file should be clear enough that editing it is straightforward.  A negative value for a weight means that the line will be weighted by its observed intensity; a positive value means that the line will be weighted by the given value.  Any weights not specified in FILE take their values from the default configuration file.

.TP
.B \-\-citation
Prints out the bibliographic details of the paper to cite if you use \fBneat\fR in your research

.TP
.B \-e, \-\-extinction\-law [VALUE]
Extinction law to use for dereddening.

Valid values are:
 \fBHow\fR: Galactic law of Howarth (1983, MNRAS, 203, 301)
 \fBCCM\fR: Galactic law of Cardelli, Clayton, Mathis (1989, ApJ, 345, 245)
 \fBFitz\fR: Galactic law of Fitzpatrick & Massa (1990, ApJS, 72, 163)
 \fBLMC\fR: LMC law of Howarth (1983, MNRAS, 203, 301)
 \fBSMC\fR: SMC law of Prevot et al. (984, A&A, 132, 389)

Default: How

.TP
.B \-hb, \-\-hbeta\-flux [VALUE]
Flux of H beta. If your line list does not include H beta, use this option to provide a value, so that NEAT can analyse the line list and compute abundances. If this command line option is used when H beta is already in the line list, the command line option will override the value in the line list.

.TP
.B \-he, \-\-helium\-data [VALUE]
The atomic data to use for He I abundances

Valid values are:
 \fBS96\fR: Smits, 1996, MNRAS, 278, 683
 \fBP12\fR: Porter et al., 2012, MNRAS, 425, 28

Default: P12

.TP
.B \-i, \-\-input [FILENAME]
The line list to be analysed.  Plain text files containing two, three or four columns of numbers can be understood by \fBneat\fR.  Full details of the input format are given in the "methodology" section below.

.TP
.B \-icf, \-\-ionisation\-correction\-scheme [VALUE]
The ICF scheme to be used to correct for unseen ions

Valid values are:
 \fBKB94\fR: Kingsburgh & Barlow (1994, MNRAS, 271, 257)
 \fBPT92\fR: Peimbert, Torres\-Peimbert & Ruiz (1992, RMxAA, 24, 155)
 \fBDI14\fR: Delgado\-Inglada, Morisset & Stasinska (2014, MNRAS, 440, 536)
 \fBDI14mod\fR: DI14, but using the classical ICF for nitrogen (N/O=N+/O+)

Default: DI14.  The KB94 ICF implemented in \fBneat\fR additionally contains a correction for chlorine, from Liu et al., 2000, MNRAS, 312, 585.

.TP
.B \-id, \-\-identify
This option triggers an algorithm which attempts to convert observed wavelengths into rest wavelengths. If the option is not present, \fBneat\fR assumes that the input line list already has rest wavelengths in the first column.

.TP
.B \-idc, \-\-identify\-confirm
Does the same as \-\-identify but without stopping for user confirmation. Useful when reanalysing a line list where you already used \-\-identify and were happy with the result.

.TP
.B \-n, \-\-n\-iterations [VALUE]
Number of iterations. Default: 1

.TP
.B \-nelow, \-nemed, \-nehigh, \-telow, \-temed, \-tehigh [VALUE]
By default, electron densities and temperatures are calculated from the available diagnostics, but their values can be overridden with these commands.  The required units are cm\-3 for densities and K for temperatures.

.TP
.B \-\-no\-omp
This option disables parallelisation so that if uncertainties are being calculated, only one thread is used.

.TP
.B \-rp
When calculating Monte Carlo uncertainties, \fBneat\fR's default behaviour is to assume that all uncertainties are Gaussian.  If \-rp is specified, it will compensate for the upward bias affecting weak lines described by Rola and Pelat (1994), assuming log normal probability distributions for weaker lines.
Until version 1.8, the default behaviour was the opposite; log normal probability distributions were assumed unless \fB-norp\fR was specified.  This was changed after our discovery that the effect described by Rola and Pelat only occurred under extremely specific conditions: see Wesson et al. 2016 for details.

.TP
.B \-sr, \-\-subtract\-recombination
The recombination contribution to some important diagnostic collisionally excited lines is always calculated but, by default, is not subtracted. This option causes the subtraction to be carried out. Note that the code takes roughly twice as long to run if this is enabled, as the temperature and abundance calculations need to be repeated following the subtraction.

.TP
.B \-u, \-\-uncertainties
Calculate uncertainties, using 10,000 iterations.  If this option is specified, the \-n option will be ignored

.TP
.B \-v, \-\-verbosity [VALUE]
Amount of output to write for each derived quantity. This option has no effect unless \-n or \-u is specified.

Valid values are:
 1: write out summary files, binned results and complete results
 2: write out summary files and binned results
 3: write out summary files only

Default: 3

.SH USING NEAT
.SS Input file format

\fBneat\fR requires as input a plain text file containing a list of emission line fluxes. The file can contain two, three or four columns. If two columns are found, the code assumes they contain the the laboratory wavelength of the line (λ0) and its flux (F). Three columns are assumed to be λ0, F, and the uncertainty on F (ΔF). Four columns are assumed to be the observed wavelength of the line, λobs, λ0, F, and ΔF. The rest wavelengths should correspond exactly to those listed in the file /usr/share/neat/complete_line_list. The flux column can use any units, and the uncertainty on the flux should be given in the same units. Examples can be found in the /usr/share/neat/examples/ directory.

.SS Rest wavelengths
\fBneat\fR identifies lines by comparing the quoted wavelength to its list of rest wavelengths. However, rest wavelengths of lines can differ by up to a few tenths of an Angstrom depending on the source. Making sure that \fBneat\fR is identifying your lines properly is probably the most important step in using the code, and misidentifications are the first thing to suspect if you get unexpected results. To assist with preparing the line list, the command line option \fB-id\fR can be used. This applies a very simple algorithm to the input line list to determine their rest wavelengths, which works as follows:

 1. A reference list of 10 rest wavelengths of common bright emission lines is compared to the wavelengths of the 20 brightest lines in the observed spectrum.
 2. Close matches are identified and the mean offset between observed and rest wavelengths is calculated.
 3. The shift is applied, and then an RMS scatter between shifted and rest wavelengths is calculated.
 4. This RMS scatter is then used as a tolerance to assign line IDs based on close coincidences between shifted observed wavelengths and the full catalogue of rest wavelengths listed in utilities/complete_line_list

The routine is not intended to provide 100% accuracy and one should always check very carefully whether the lines are properly identified, particularly in the case of high resolution spectra.

.SS Line blends

In low resolution spectra, lines of comparable intensity may be blended into a single feature. These can be indicated with an asterisk instead of a flux in the input line list. Currently, \fBneat\fR has only limited capabilities for dealing with blends: lines marked as blends are not used in any abundance calculations, and apart from a few cases, it assumes that all other line fluxes represent unblended or deblended intensities. The exceptions are some collisionally excited lines which are frequently blended, such as the [O II] lines at 3727/3729Å. In these cases the blended flux can be given with the mean wavelength of the blend, and the code will treat it properly. These instances are indicated in the utilities/complete_line_list file by a "b" after the ion name.

.SS Uncertainties

The uncertainty column of the input file is of crucial importance if you want to estimate uncertainties on the results you derive. Points to bear in mind are that the more realistic your estimate of the line flux measurement uncertainties, the more realistic the estimate of the uncertainties on the results will be, and that in all cases, the final reported uncertainties are a lower limit to the actual uncertainty on the results, because they account only for the propagation of the statistical errors on the line fluxes and not on sources of systematic uncertainty.

In some cases you may not need or wish to propagate uncertainties. In this case you can run just one iteration of the code, and the uncertainty values are ignored if present.

.SH RUNNING THE CODE
Assuming you have a line list prepared as above, you can now run the code. In line with our philosophy that \fBneat\fR should be as simple and objective as possible, this should be extremely straightforward. To use the code in its simplest form on one of the example linelists, you would type

 % cp /usr/share/neat/examples/ngc6543_3cols.dat .
 % neat \-i ngc6543_3cols.dat

This would run a single iteration of the code, not propagating uncertainties. You'll see some logging output to the terminal, and the calculated results will have been written to the file ngc6543_3cols.dat_results. If this is all you need, then the job's done and you can write a paper now.

Your results will be enhanced greatly, though, if you can estimate the uncertainty associated with them. To do this, invoke the code as follows:

 % neat \-i ngc6543_3cols.dat \-u

The \-u switch causes the code to run 10,000 times. In each iteration, the line flux is drawn from a normal distribution with a mean of the quoted flux and a standard deviation of the quoted uncertainty.  By repeating this randomisation process lots of times, you build up a realistic picture of the uncertainties associated with the derived quantities. The more iterations you run, the more accurate the results; 10,000 is a sensible number to achieve well sampled probability distributions. If you want to run a different number of iterations for any reason, you can use the \-n command line option to specify your preferred value

If the \fB-rp\fR option is specified, then for lines with a signal to noise ratio of less than 6, the line flux is drawn from a log-normal distribution which becomes more skewed the lower the signal to noise ratio is. This corrects the low SNR lines for the upward bias in their measurement described by Rola & Pelat (1994). The full procedure is described in Wesson et al. (2012).  However, use of this option is no longer recommended as the bias is highly dependent on the fitting procedure \- see Wesson et al. (2016).

.SS Error codes

If the code does not complete the analysis, it will return one of the following exit codes:

 100  no input file specified
 101  input file does not exist
 102  config file does not exist
 103  invalid number of iterations
 104  invalid output format requested
 105  could not read input line list
 106  could not read LINES extension in FITS input
 107  error reading line list from FITS
 108  error creating FITS output
 109  invalid configuration file

 200  invalid temperature or density
 201  no Hbeta found
 202  unknown ion
 203  line identifications rejected

In the case of a FITS handling error, it will also return the CFITSIO exit code.

.SH METHODOLOGY

.SS Extinction correction

The code corrects for interstellar reddening using the ratios of the Hα, Hβ, Hγ and Hδ lines. Intrinsic ratios of the lines are first calculated assuming a temperature of 10,000K and a density of 1000cm-3. The line list is then dereddened, and temperatures and densities are then calculated as described below. The temperatures and densities are then used to recalculate the intrinsic Balmer line ratios, and the original line list is then dereddened using this value.

.SS Temperatures and densities

\fBneat\fR determines temperatures, densities and abundances by dividing emission lines into low (ionisation potential <20eV), medium (20eV<IP<45eV) and high excitation (IP>45eV) lines. In each zone, the diagnostics are calculated as follows:

 1. A temperature of 10000K is initially assumed, and the density is then calculated from the line ratios relevant to the zone.
 2. The temperature is then calculated from the temperature diagnostic line ratios, using the derived density.
 3. The density is recalculated using the appropriately weighted average of the temperature diagnostics.
 4. The temperature is recalculated using this density.

This iterative procedure is carried out successively for low-, medium- and high-ionization zones, and in each case if no diagnostics are available, the temperature and/or density will be taken to be that derived for the previous zone. Temperatures and densities for each zone can also be specified on the command line with the \fB-telow, \-temed, \-tehigh\fR and \fB-nelow, \-nemed, \-nehigh\fR options.

\fBneat\fR also calculates a number of diagnostics from recombination line diagnostics. These are:

 1. The Balmer jump temperature is calculated using equation 3 of Liu et al. (2001)
 2. The Paschen jump temperature is calculated using equation 7 of Fang et al. (2011)
 3. A density is derived from the Balmer and Paschen decrements if any lines from H10-H25 or P10-P25 are observed. Their ratios relative to Hβ are compared to theoretical ratios from Storey & Hummer (1995), and a density for each line calculated by linear interpolation. The final density is calculated as the weighted average of all the densities.
 4. Temperatures are estimated from helium line ratios, using equations derived from fits to tabulated values of 5876/4471 and 6678/4471. The tables are calculated at ne=5000cm-3 only. We plan to improve this calculation in future releases.
 5. OII recombination line ratios are used to derive a temperature and density, using atomic data calculations from Storey et al. (2017). Values are found by linearly interpolating the logarithmic values.
 6. Recomination line contributions to CELs of N+, O+ and O2+ are estimated using equations 1-3 of Liu et al. (2000).

These recombination line diagnostics are not used in abundance calculations. By default, the recombination contribution to CELs is reported but not subtracted. The command line option \fB\-\-subtract\-recombination\fR can be used if the subtraction is required - this requires an additional loop within the code which makes it run roughly half as fast as when the subtraction is not carried out.

.SS Ionic abundances

Ionic abundances are calculated from collisionally excited lines (CELs) using the temperature and density appropriate to their ionization potential. Where several lines from a given ion are present, the ionic abundance adopted is a weighted average of the abundances from each ion.

Recombination lines (RLs) are also used to derive ionic abundances for helium and heavier elements. The method by which the helium abundance is determined depends on the atomic data set being used; \fBneat\fR includes atomic data from Smits (1996) and from Porter et al. (2012, 2013). The Smits data is given for seven temperatures between 312.5K and 20000K, and for densities of 1e2, 1e4 and 1e6 cm-3; we fitted fourth order polynomials to the coefficient for each line at each density. \fBneat\fR then calculates the emissivities for each density using these equations, and interpolates logarithmically to the correct density.

For the Porter et al. data, the emissivities are tabulated between 5000 and 25000K, and for densities up to 1e14cm-3. \fBneat\fR interpolates logarithmically in temperature and density between the tabulated values to determine the appropriate emissivity.

In deep spectra, many more RLs may be available than CELs. The code calculates the ionic abundance from each individual RL intensity using the atomic data listed in Table 1 of Wesson et al. (2012). Then, to determine the ionic abundance to adopt, it first derives an ionic abundance for each individual multiplet from the multiplet’s co-added intensity, and then averages the abundances derived for each multiplet to obtain the ionic abundance used in subsequent calculations.

The weakness of recombination lines means that care must be taken in deriving abundances from them. \fBneat\fR applies some simple checks to the N2+/H+ and O2+/H+ abundances it derives to see if they are reliable: they will be flagged as unreliable in the _results file if any of the following apply:

 1. only one multiplet is detected
 2. the highest abundance calculated for a multiplet exceeds the lowest by a factor greater than 3.0
 3. (O2+ only) no lines are detected from either the V1 or V10 multiplets

If either abundance is flagged as unreliable, an abundance discrepancy factor for the relevant ion and element is not calculated. Weights may be set to exclude any unreliable multiplets so that a reliable abundance and ADF can be calculated.

.SS Total abundances

Total elemental abundances are estimated using the ionisation correction scheme selected from Kingsburgh and Barlow (1994), Peimbert, Torres-Peimbert and Ruiz (1992), or Delgado-Inglada et al. (2014). Total oxygen abundances estimated from several strong line methods are also reported.

Where ionic or total abundances are available from both collisionally excited lines and recombination lines, the code calculates the measured discrepancy between the two values.

.SH OUTPUTS

The code prints some logging messages to the terminal, so that you can see which iteration it is on, and if anything has gone wrong. Starting from version 2.3, outputs are created in FITS format by default. If the input file was in FITS format (as generated by v2.0 or later of \fBalfa\fR) then the results of the analysis will be written to the same file, updating the LINES extension and creating a RESULTS extension. If the input was plain text, then a new FITS file will be created, in the same directory as the input file. \fBfv\fR is a convenient way to view the tables.


source/neat.f90
! NEAT, the nebular abundance analysis tool
! (C) 2006- Roger Wesson, Dave Stock, Peter Scicluna
! NEAT incorporates aspects of several codes developed over decades at UCL:
! equib, by I. Howarth and S. Adams, updated significantly by B. Ercolano
! MIDAS scripts written by X-W. Liu for calculating recombination line abundances
! It also uses the quicksort algorithm as implemented in F90 by Alberto Ramos

! This program is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.

! This program is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.

! You should have received a copy of the GNU General Public License
! along with this program.  If not, see <http://www.gnu.org/licenses/>.

program neat

        use mod_types
        use mod_globals
        use mod_resultarrays
        use mod_extinction
        use mod_quicksort
        use mod_abundIO
        use mod_atomicdata
        use mod_atomic_read
        use mod_recombination_lines
        use mod_linefinder
        use mod_hydrogen
        use mod_helium
        use mod_oii_diagnostics
        use mod_weights
        use mod_functions
        use mod_commandline
        use mod_output

        implicit none

        character :: switch_ext !switch for extinction laws
        character :: switch_he  !switch for helium atomic data
        character :: switch_icf !switch for which ICF scheme to use
        integer :: I, runs !runs = number of runs for randomiser

!file reading variables

        logical :: identifylines,identifyconfirm
        type(line),dimension(:), allocatable :: linelist
        type(line),dimension(:), allocatable :: linelist_original
        type(line),dimension(:,:), allocatable :: all_linelists
        integer :: listlength, ncols, strpos
        real(kind=dp) :: normalise

!results and result processing

        type(resultarray), dimension(:), allocatable :: all_results
        type(resultarray), dimension(1) :: iteration_result

!atomic data

        character(len=10), dimension(24) :: ionlist !list of ion names
        integer :: iion !# of ions in zones.dat
        integer :: maxlevs,maxtemps
        type(atomic_data),allocatable :: atomicdata(:)
        real(kind=dp), dimension(:,:,:), allocatable :: heidata

!line indexing arrays - these contain the location within the main array of the lines of interest, plus levels and zones for CELs

        integer, dimension(3:40) :: H_Balmer
        integer, dimension(4:39) :: H_Paschen
        integer, dimension(44) :: HeI_lines
        integer, dimension(20,2:6) :: HeII_lines
        type(cel), dimension(88) :: ILs !todo: work out why this becomes undefined on entry if its shape is assumed

!extinction

        logical :: calculate_extinction
        real(kind=dp) :: meanextinction, R

!RP effect switch

        logical :: norp

!subtract recombination contribution to auroral lines switch - 1=no subtraction, 2=subtraction

        integer :: subtract_recombination

!diagnostics

        type(weightingarray) :: weights

!binning and uncertainties

        integer :: verbosity,nbins,nperbin

!OpenMP

        integer :: omp_get_num_threads
        logical :: noomp

!diagnostic array

        type(diagnostic_array) :: diagnostics

!manually set hbeta flux

        real(kind=dp) :: hbetaflux

! default values

        runs=1
        switch_ext="S" !Howarth 1983 Galactic law
        switch_he="P"  !Porter 2012 data
        switch_icf="D" !DI14 ICF
        filename=""
        outputdirectory=""
        meanextinction=0.D0
        diagnostics%lowtemp=0.D0
        diagnostics%lowdens=0.D0
        diagnostics%medtemp=0.D0
        diagnostics%meddens=0.D0
        diagnostics%hightemp=0.D0
        diagnostics%highdens=0.D0
        verbosity=3
        R=3.1
        identifylines=.false.
        identifyconfirm=.false.
        nbins=25
        normalise = 0.d0
        norp=.true.
        calculate_extinction=.true.
        subtract_recombination=1
        configfile=""
        defaultconfigfile=trim(PREFIX)//"/share/neat/default.cfg"
        outputformat="fits"
        hbetaflux=0.d0
        noomp=.false.

        iion=24 ! number of ions for which we can calculate abundances
                ! todo: calculate this automatically
                ! and replace it in calls to subroutines with getting the array size within the subroutine

        print *,"NEAT, the Nebular Empirical Analysis Tool"
        print *,"version ",VERSION
        print *
        print *,gettime(),"starting code"

        call readcommandline(runs,switch_ext,switch_he,switch_icf,meanextinction,diagnostics,verbosity,R,identifylines,identifyconfirm,nbins,norp,calculate_extinction,subtract_recombination,configfile,nperbin,hbetaflux,noomp)

! set up filenames. if input file is FITS, output will be written to it. Output will be put in same directory as input by default.
! todo: allow non-FITS to be requested

        write (outputfilename,"(A)") filename(index(filename,"/",back=.true.)+1:len(trim(filename)))

        if (trim(outputdirectory).eq."") then
          outputdirectory=trim(filename(1:index(filename,"/",.true.)))
        endif

        if (trim(outputdirectory).eq."") then
          outputdirectory="./"
        endif

        outputfilename=trim(outputdirectory)//"/"//trim(outputfilename)

! read in the line list, allocate original linelist array

        strpos=index(filename,".",.true.)+1
        if (trim(filename(strpos:len(filename))).eq."fit" .or. trim(filename(strpos:len(filename))).eq."fits".or.trim(filename(strpos:len(filename))).eq."FIT".or.trim(filename(strpos:len(filename))).eq."FITS") then
          call read_fits_linelist(linelist,listlength,ncols)
        else
          call read_text_linelist(linelist,listlength,ncols,runs)
        endif

! add hbeta if manually specified
        if (hbetaflux.gt.0d0) then
          if (minval(abs(linelist(:)%wavelength - 4861.33)) .lt. 0.001) then
! overwrite, warn
            linelist(minloc(abs(linelist(:)%wavelength - 4861.33),1))%intensity=hbetaflux
            print *,gettime(),"warning: manually specified Hbeta flux overrides value already in line list"
          else
! create array one element longer
            listlength=listlength+1
            allocate (linelist_original(listlength+1))

! copy linelist to it
            linelist_original = linelist

! deallocate linelist, reallocate one element longer
            deallocate(linelist)
            allocate(linelist(listlength))

! copy original lines back and deallocate
            linelist(1:listlength)=linelist_original
            deallocate(linelist_original)

! add hbeta
            linelist(listlength)%intensity = hbetaflux
            linelist(listlength)%abundance = 0.D0
            linelist(listlength)%weight = 0.d0
            linelist(listlength)%freq=0d0
            linelist(listlength)%wavelength=4861.33
            linelist(listlength)%wavelength_observed=4861.33
            linelist(listlength)%int_dered=0d0
            linelist(listlength)%int_err=0d0
            linelist(listlength)%blend_intensity=0.d0
            linelist(listlength)%blend_int_err=0d0
            linelist(listlength)%zone='    '
            linelist(listlength)%transition='                    '
            linelist(listlength)%location=0
            linelist(listlength)%ion='H 4       '
            linelist(listlength)%multiplet='H4          '
            linelist(listlength)%lowerterm='2p+ 2P*     '
            linelist(listlength)%upperterm='4d+ 2D      '
            linelist(listlength)%g1=8
            linelist(listlength)%g2=32

          endif
        endif

        allocate (linelist_original(listlength))

! run line identifier if required

        if (identifylines) then
                call linefinder(linelist, listlength, identifyconfirm)
        endif

! normalise to Hbeta

        if (minval(abs(linelist(:)%wavelength - 4861.33)) .lt. 0.001) then
          normalise = 100.d0/linelist(minloc(abs(linelist(:)%wavelength - 4861.33),1))%intensity
        else
          print *,gettime(),"error: no H beta detected. No further analysis possible"
          print *,gettime(),"       -hb / --hbeta-flux command line option can be used to give a value"
          call exit(201)
        endif

        linelist%intensity = linelist%intensity * normalise
        linelist%int_err = linelist%int_err * normalise
        linelist%blend_intensity = linelist%blend_intensity * normalise
        linelist%blend_int_err = linelist%blend_int_err * normalise

! check for and remove negative line fluxes and uncertainties

        if (minval(linelist%intensity) .lt. 0.) then
          where (linelist%intensity .lt. 0)
            linelist%intensity = 0.D0
          endwhere
        print *,gettime(),"warning: negative line fluxes set to zero"
        endif

        if (minval(linelist%int_err) .lt. 0.) then
          where (linelist%int_err .lt. 0)
            linelist%int_err = abs(linelist%int_err)
          endwhere
        print *,gettime(),"warning: negative flux uncertainties converted to positive"
        endif

        print *

! select reddening law, get flambda from functions if one of the five, otherwise read from file

        if (switch_ext == "S") then
                print *,gettime(),"using Howarth (1983) galactic extinction law"
        elseif (switch_ext == "H") then
                print *,gettime(),"using Howarth (1983) LMC extinction law"
        elseif (switch_ext == "C") then
                print *,gettime(),"using CCM (1989) galactic extinction law"
        elseif (switch_ext == "P") then
                print *,gettime(),"using Prevot et al. (1984) SMC extinction law"
        elseif (switch_ext == "F") then
                print *,gettime(),"using Fitzpatrick (1990) galactic extinction law"
        endif

        call get_flambda(linelist, switch_ext, R)

        if (switch_he == "S") then
                print *,gettime(),"using Smits (1996) He I emissivities"
        else
                print *,gettime(),"using Porter et al. (2012) He I emissivities"
        endif

        if (switch_ICF == "K") then
                print *,gettime(),"using Kingsburgh & Barlow (1994) ICF"
        elseif (switch_ICF == "D") then
                print *,gettime(),"using Delgado-Inglada et al. (2014) ICF"
        elseif (switch_ICF == "E") then
                print *,gettime(),"using Delgado-Inglada et al. (2014) ICF with classical icf(N)"
        else
                print *,gettime(),"using Peimbert et al. (1992) ICF"
        endif

        print *

! read the CEL data

        call read_celdata(ILs,ionlist)

! do line identifying here

        !first find H & He lines, and CELs
        call get_H(H_Balmer, H_Paschen, linelist)
        call get_Hei(Hei_lines, linelist)
        call get_Heii(Heii_lines, linelist)
        call get_cels(ILs,linelist)

! read in the weights to be used in the analysis.

        print *,gettime(),"reading in abundance analysis weights from ",trim(defaultconfigfile)
        call setweights(defaultconfigfile,weights,linelist,ILs,H_Balmer,H_Paschen,HeII_lines)
        if (configfile .ne. "") then
          print *,gettime(),"reading in abundance analysis weights from ",trim(configfile)
          print *,gettime(),"(defaults still apply for any weights not specified in this file)"
          call setweights(configfile,weights,linelist,ILs,H_Balmer,H_Paschen,HeII_lines)
        endif

!CELs read, can now read in atomic data

        print *,gettime(),"reading atomic data from ",trim(PREFIX),"/share/neat"
        allocate(atomicdata(iion))
        do I=1,iion
            atomicdata(I)%ion = ionlist(I)
            call read_atomic_data(atomicdata(I))
        enddo

!read ORL data
        call read_orl_data

!read hydrogen emissivities

        print *,gettime(),"reading H emissivities"
        call read_hydrogen

!read helium emissivities

        print *,gettime(),"reading He I emissivities"
        if (switch_he .eq. "P") then
          allocate(heidata(21,14,44))
          heidata = 0.D0
          call read_porter(heidata)
        elseif (switch_he .eq. "S") then
          allocate(heidata(3,6,44))
          heidata = 0.D0
          call read_smits(heidata)
        endif

        print *,gettime(),"reading He II emissivities"
        call read_heii

!read oii data for diagnostics

        print *,gettime(),"reading OII diagnostics data"
        call read_oii_diagnosticdata
        call read_oii_s2017

!find maximum #levels and temperatures - pass to equib to reduce footprint

        maxlevs = maxval(atomicdata%nlevs)
        maxtemps = maxval(atomicdata%ntemps)

!omp setup
        if (noomp) call omp_set_num_threads(1)

!now check number of iterations.  If 1, line list is fine as is.  If more than one, randomize the fluxes

        allocate(all_results(runs))
        allocate(all_linelists(size(linelist),runs))

        if (runs == 1)then !calculates abundances without uncertainties
                print *
                print *,gettime(),"doing abundance calculations"

                call abundances(linelist, listlength, iteration_result, meanextinction, calculate_extinction, ILs, diagnostics,iion,atomicdata,maxlevs,maxtemps, heidata, switch_he, switch_icf, H_Balmer, H_Paschen, HeI_lines, HeII_lines, weights,subtract_recombination)

                all_results(1)=iteration_result(1) ! copy the iteration result to all_results to simplify the writing out of results later
                all_linelists(:,1)=linelist

                print *,gettime(),"finished abundance calculations"
        else if (runs .gt. 1)then

!save unrandomised line list

                linelist_original = linelist

                call init_random_seed()!sets seed for randomiser

                !main loop
!$OMP PARALLEL default(firstprivate) shared(all_linelists,listlength,norp,calculate_extinction,ILs,diagnostics,iion,atomicdata,maxlevs,maxtemps,switch_he,switch_icf,all_results,subtract_recombination)
!$OMP MASTER

                print *
                if (omp_get_num_threads().gt.1) then
                  print "(X,A10,X,A,I2,A)",gettime(),"starting Monte Carlo calculations, using ",omp_get_num_threads()," processors"
                else
                  print "(X,A10,X,A)",gettime(),"starting Monte Carlo calculations"
                endif
                print *,gettime(),"completed ",0,"%"
!$OMP END MASTER
!$OMP DO schedule(dynamic)
                do I=1,runs

                         if ( (10.0*dble(i)/dble(runs)) == int(10*i/runs) ) print *,gettime(),"completed ",100*i/runs,"%"
!                        print*, "iteration ", i, "of", runs

                        call randomizer(linelist, listlength, norp)
                        call abundances(linelist, listlength, iteration_result, meanextinction, calculate_extinction, ILs, diagnostics,iion,atomicdata,maxlevs,maxtemps, heidata, switch_he, switch_icf, H_Balmer, H_Paschen, HeI_lines, HeII_lines, weights, subtract_recombination)

                        !store all line and derived quantity in arrays
                        all_linelists(:,i)=linelist
                        all_results(i)=iteration_result(1)
                        !restore the unrandomized line list ready for the next iteration
                        linelist = linelist_original
                enddo
!$OMP END DO
!$OMP END PARALLEL
        else
                print*, gettime(),"error: I didn't want to be a barber anyway. I wanted to be... a lumberjack!   Also, a positive number of runs helps.."
                call exit(103)
        endif

!now write out the line list and results files

        if (outputformat.eq."fits") then
          call write_fits(runs,listlength,ncols,all_linelists,all_results,nbins)
        elseif (outputformat.eq."text") then
          call write_output(runs,listlength,ncols,all_linelists,all_results,verbosity,nbins,subtract_recombination)
        endif

! deallocate
        deallocate(all_linelists)
        deallocate(all_results)
        deallocate(linelist)
        deallocate(linelist_original)
        deallocate(hidata)
        deallocate(heidata)
        deallocate(heiidata)
        deallocate(atomicdata)
        deallocate(ciirls)
        deallocate(niirls)
        deallocate(oiirls)
        deallocate(neiirls)
        deallocate(xiiirls)
        deallocate(hidatatemperatures)
        deallocate(heiidatatemperatures)

        print *
        print *,gettime(),"all done"
        print *

end program
